File: create.js
const AWS = require('aws-sdk');
const dynamoDB = new AWS.DynamoDB.DocumentClient();

// 配置
const TABLE_NAME = process.env.DYNAMODB_TABLE;
const BASE_URL = process.env.BASE_URL;
const SHORT_CODE_LENGTH = 6;

/**
 * 生成指定长度的随机短码
 * @param {number} length - 短码长度
 * @returns {string} - 随机短码
 */
function generateShortCode(length) {
  const charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';

  for (let i = 0; i < length; i++) {
    const randomIndex = Math.floor(Math.random() * charset.length);
    result += charset[randomIndex];
  }

  return result;
}

/**
 * 在DynamoDB中存储URL映射
 * @param {string} shortCode - 生成的短码
 * @param {string} originalUrl - 要重定向到的原始URL
 * @returns {Promise} - DynamoDB操作promise
 */
async function storeUrlMapping(shortCode, originalUrl) {
  const timestamp = Math.floor(Date.now() / 1000);

  const params = {
    TableName: TABLE_NAME,
    Item: {
      short_code: shortCode,
      original_url: originalUrl,
      created_at: timestamp,
      visit_count: 0
    },
    // 确保短码不存在
    ConditionExpression: 'attribute_not_exists(short_code)'
  };

  try {
    await dynamoDB.put(params).promise();
    return true;
  } catch (error) {
    // 如果短码已存在，返回false
    if (error.code === 'ConditionalCheckFailedException') {
      return false;
    }
    throw error;
  }
}

/**
 * 用于创建短链接的Lambda处理函数
 */
exports.handler = async (event) => {
  try {
    // 解析请求体
    const body = JSON.parse(event.body);

    // 验证输入
    if (!body.url) {
      return {
        statusCode: 400,
        headers: {
          'Content-Type': 'application/json',
          "Access-Control-Allow-Origin": "http://" + process.env.FRONTEND_DOMAIN,
          'Access-Control-Allow-Methods': 'POST, OPTIONS',
          'Access-Control-Allow-Headers': 'Content-Type'
        },
        body: JSON.stringify({ error: '需要提供URL' })
      };
    }

    const originalUrl = body.url;

    // 尝试最多3次生成唯一短码
    let shortCode;
    let isStored = false;
    let attempts = 0;

    while (!isStored && attempts < 3) {
      shortCode = generateShortCode(SHORT_CODE_LENGTH);
      isStored = await storeUrlMapping(shortCode, originalUrl);
      attempts++;
    }

    if (!isStored) {
      return {
        statusCode: 500,
        headers: {
          'Content-Type': 'application/json',
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Methods': 'POST, OPTIONS',
          'Access-Control-Allow-Headers': 'Content-Type'
        },
        body: JSON.stringify({ error: '无法生成唯一短码' })
      };
    }

    // 构建短链接
    const shortUrl = `${BASE_URL}/${shortCode}`;

    // 返回短链接
    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      },
      body: JSON.stringify({
        originalUrl,
        shortCode,
        shortUrl
      })
    };
  } catch (error) {
    console.error('错误:', error);

    return {
      statusCode: 500,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
      },
      body: JSON.stringify({ error: '内部服务器错误' })
    };
  }
};

File: create.py
import json
import boto3
import os
import random
import string
import time
from typing import Dict, Any, Optional
import logging
from botocore.exceptions import ClientError

# Setup logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Configuration
TABLE_NAME = os.environ['DYNAMODB_TABLE']
BASE_URL = os.environ['BASE_URL']
SHORT_CODE_LENGTH = 6

# AWS clients
dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table(TABLE_NAME)

def generate_short_code(length: int) -> str:
    """
    Generate a random short code of specified length
    
    Args:
        length (int): Length of the short code
        
    Returns:
        str: Random short code
    """
    charset = string.ascii_letters + string.digits
    return ''.join(random.choice(charset) for _ in range(length))

def store_url_mapping(short_code: str, original_url: str) -> bool:
    """
    Store URL mapping in DynamoDB
    
    Args:
        short_code (str): Generated short code
        original_url (str): Original URL to redirect to
        
    Returns:
        bool: True if stored successfully, False if short code already exists
    """
    timestamp = int(time.time())
    
    try:
        table.put_item(
            Item={
                'short_code': short_code,
                'original_url': original_url,
                'created_at': timestamp,
                'visit_count': 0
            },
            # Ensure short code doesn't exist
            ConditionExpression='attribute_not_exists(short_code)'
        )
        return True
    except ClientError as e:
        # If short code already exists, return False
        if e.response['Error']['Code'] == 'ConditionalCheckFailedException':
            return False
        raise e

def create_cors_headers(origin: Optional[str] = None) -> Dict[str, str]:
    """
    Create CORS headers for the response
    
    Args:
        origin (Optional[str]): Specific origin to allow, defaults to wildcard
        
    Returns:
        Dict[str, str]: CORS headers
    """
    allowed_origin = origin if origin else '*'
    if 'FRONTEND_DOMAIN' in os.environ and origin:
        allowed_origin = f"http://{os.environ['FRONTEND_DOMAIN']}"
    
    return {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': allowed_origin,
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type'
    }

def lambda_handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler for creating short links
    
    Args:
        event (Dict[str, Any]): Lambda event containing request data
        context (Any): Lambda context object
        
    Returns:
        Dict[str, Any]: HTTP response with short URL or error
    """
    try:
        # Parse request body
        body = json.loads(event.get('body', '{}'))
        
        # Validate input
        if not body.get('url'):
            return {
                'statusCode': 400,
                'headers': create_cors_headers(),
                'body': json.dumps({'error': 'URL is required'})
            }
        
        original_url = body['url']
        
        # Attempt to generate unique short code up to 3 times
        short_code = None
        is_stored = False
        attempts = 0
        
        while not is_stored and attempts < 3:
            short_code = generate_short_code(SHORT_CODE_LENGTH)
            is_stored = store_url_mapping(short_code, original_url)
            attempts += 1
        
        if not is_stored:
            return {
                'statusCode': 500,
                'headers': create_cors_headers(),
                'body': json.dumps({'error': 'Unable to generate unique short code'})
            }
        
        # Build short URL
        short_url = f"{BASE_URL}/{short_code}"
        
        # Return short URL
        return {
            'statusCode': 200,
            'headers': create_cors_headers(),
            'body': json.dumps({
                'originalUrl': original_url,
                'shortCode': short_code,
                'shortUrl': short_url
            })
        }
        
    except json.JSONDecodeError:
        return {
            'statusCode': 400,
            'headers': create_cors_headers(),
            'body': json.dumps({'error': 'Invalid JSON in request body'})
        }
    except Exception as error:
        logger.error(f"Error: {str(error)}")
        
        return {
            'statusCode': 500,
            'headers': create_cors_headers(),
            'body': json.dumps({'error': 'Internal server error'})
        }